crates/bitwarden/src/tool/generators/passphrase.rs

```diff
@@ -0,0 +1,161 @@
+use crate::{error::Result, wordlist::EFF_LONG_WORD_LIST};
+use rand::{seq::SliceRandom, Rng, RngCore};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+/// Passphrase generator request.
+///
+/// The default separator is `-` and default number of words is 3.
+#[derive(Serialize, Deserialize, Debug, JsonSchema, Default)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+#[cfg_attr(feature = "mobile", derive(uniffi::Record))]
+pub struct PassphraseGeneratorRequest {
+    pub num_words: Option<u8>,
+    pub word_separator: Option<String>,
+    pub capitalize: Option<bool>,
+    pub include_number: Option<bool>,
+}
+
+const DEFAULT_PASSPHRASE_NUM_WORDS: u8 = 3;
+const DEFAULT_PASSPHRASE_SEPARATOR: char = ' ';
+
+pub(super) fn passphrase(input: PassphraseGeneratorRequest) -> Result<String> {
+    passphrase_with_rng(rand::thread_rng(), input)
+}
+
+fn passphrase_with_rng(mut rng: impl RngCore, input: PassphraseGeneratorRequest) -> Result<String> {
+    let num_words = input.num_words.unwrap_or(DEFAULT_PASSPHRASE_NUM_WORDS);
+    let separator = input
+        .word_separator
+        .and_then(|s| s.chars().next())
+        .unwrap_or(DEFAULT_PASSPHRASE_SEPARATOR);
+    let capitalize = input.capitalize.unwrap_or(false);
+    let include_number = input.include_number.unwrap_or(false);
+
+    let mut passphrase_words = gen_words(&mut rng, num_words);
+    if include_number {
+        include_number_in_words(&mut rng, &mut passphrase_words);
+    }
+    if capitalize {
+        capitalize_words(&mut passphrase_words);
+    }
+    Ok(passphrase_words.join(&separator.to_string()))
+}
+
+fn gen_words(mut rng: impl RngCore, num_words: u8) -> Vec<String> {
+    (0..num_words)
+        .map(|_| {
+            EFF_LONG_WORD_LIST
+                .choose(&mut rng)
+                .expect("slice is not empty")
+                .to_string()
+        })
+        .collect()
+}
+
+fn include_number_in_words(mut rng: impl RngCore, words: &mut [String]) {
+    let number_idx = rng.gen_range(0..words.len());
+    words[number_idx].push_str(&rng.gen_range(0..=9).to_string());
+}
+
+fn capitalize_words(words: &mut [String]) {
+    words
+        .iter_mut()
+        .for_each(|w| *w = capitalize_first_letter(w));
+}
+
+fn capitalize_first_letter(s: &str) -> String {
+    let mut c = s.chars();
+    match c.next() {
+        None => String::new(),
+        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
+    }
```

Comment:
⚠️ At first I thought direct string manipulation would be more straightforward, but forgot that unicode case conversion can grow the length of the string. I'd consider adding a comment here explaining the iterator approach solves that issue.

💭 At the same time, I still wonder whether there's a way to eliminate the match.

Reply:
Sounds good, note that there is also this option that works assuming all characters are ASCII (which for the words list would work):
```rust
fn capitalize_first_letter_ascii(s: &mut str) {
    if let Some(r) = s.get_mut(0..1) {
        r.make_ascii_uppercase();
    }
}
```
I don't think it's much more readable though

Reply:
That's where I went as well. I'd avoid it since it doesn't generalize. The main thing I like about it isn't the `if let` (though I am a fan), but that it avoids allocations.